diff --git a/filefield.module b/filefield.module
index c199990..976181b 100644
--- a/filefield.module
+++ b/filefield.module
@@ -600,11 +600,42 @@ function filefield_js($type_name, $field_name, $delta) {
 
   // Update the cached form with the new element at the right place in the form.
   if (module_exists('fieldgroup') && ($group_name = _fieldgroup_field_get_group($type_name, $field_name))) {
+    //Get the fieldgroup tree information for this field
+    $tree = _fieldgroup_get_tree($type_name);
+    $parents = array(0 => $group_name);
+
+    //Find the parent structure
+    $tmp_group_name = $group_name; //don't overwrite $group_name
+    while ($tmp_group_name = $tree[$tmp_group_name]['parents'][0]) {
+      $parents[] = $tmp_group_name;
+    }
+
+    //Put in the correct order
+    $parents = array_reverse($parents);
+    $tmp_element = $form;
+
+    //Now we need to find the field within the structure. So iterate down to its parents level.
+    foreach ($parents as $idx) {
+      if (isset($tmp_element[$idx]) && is_array($tmp_element[$idx])) {
+        $tmp_element = $tmp_element[$idx];
+      }
+    }
+
+    //the field is found now determine if this is in a multigroup
+    //using the original if/else
     if (isset($form['#multigroups']) && isset($form['#multigroups'][$group_name][$field_name])) {
-      $form_element = $form[$group_name][$delta][$field_name];
+      //multigroup structure is $tmp_element[$delta][$field_name]
+      $form_element = $tmp_element[$delta][$field_name];
     }
     else {
-      $form_element = $form[$group_name][$field_name][$delta];
+      //non-multigroup structure is $tmp_element[$field_name] or $tmp_element[$field_name][$delta]
+      $field = $tmp_element[$field_name];
+      if (isset($field[$delta])) {
+        $form_element = $field[$delta];
+      }
+      else {
+        $form_element = $field;
+      }
     }
   }

